provider/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinScriptPlugin.kt

```diff
@@ -37,14 +36,18 @@ class KotlinScriptPlugin(
 
     override fun apply(target: Any) {
         logger.debug("Applying Kotlin script to {}", target)
-        if (target !is Project) {
-            throw IllegalArgumentException("target $target was not a Project as expected")
-        }
-        target.run {
-            afterEvaluate {
-                plugins.apply(KotlinScriptBasePlugin::class.java)
-            }
+        when (target) {
```

Comment:
:x: We have already taken a decision based on the target before, could we leverage that somehow to avoid taking this decision again here?

Reply:
Actually this is the "earliest" decision point. The second one you refer too happen inside `script: (Any) -> Unit` here. We don't have the target at the point where we instantiate `KotlinScriptPlugin`.

Reply:
I moved this responsibility to `KotlinScriptPluginTarget`, much cleaner!